middlewares/auth/forward.go

```diff
@@ -52,8 +59,36 @@ func Forward(config *types.Forward, w http.ResponseWriter, r *http.Request, next
 	}
 	defer forwardResponse.Body.Close()
 
+	// Pass the forward response's body and selected headers if it
+	// didn't return a response within the range of [200, 300).
+
 	if forwardResponse.StatusCode < http.StatusOK || forwardResponse.StatusCode >= http.StatusMultipleChoices {
 		log.Debugf("Remote error %s. StatusCode: %d", config.Address, forwardResponse.StatusCode)
+
+		// Grab the location header, if any.
+
+		redirectURL, err := forwardResponse.Location()
+
+		if err != nil && err != http.ErrNoLocation {
+			log.Debugf("Error reading response location header %s. Cause: %s", config.Address, err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		// Set the location in our response if one was sent back.
+
+		if err != http.ErrNoLocation && redirectURL.String() != "" {
```

Comment:
There are three things to handle with Response.Location() return values:
1. Handle the case where a Location header was present but it did not have a parsable URL (some non-nil error).
1. Handle the case where no Location header was present (`err == http.ErrNoLocation`).
1. Handle the case where a Location header was present and it had a parsable URL.

The second case is common and not an error as far as we are concerned as any non-redirect response code would trigger this case, so it's important to ensure it passes through. Any other error was likely a non-parsable URL. With a small edit, this can be simplified in mainly the manner you describe:

```go
redirectURL, err := forwardResponse.Location()

if err != nil {
	if err != http.ErrNoLocation { // Case #1
		log.Debugf("Error reading response location header %s. Cause: %s", config.Address, err)
		w.WriteHeader(http.StatusInternalServerError)
		return
	} /* else {
		// Case #2, do nothing with http.ErrNoLocation
	} */
} else if redirectURL.String() != "" { // Case #3
	// Set the location in our response if one was sent back.
	w.Header().Add("Location", redirectURL.String())
}
```

Reply:
```go
if err != nil {
	if err != http.ErrNoLocation {
// ...
    }
} else if redirectURL.String() != "" {
// ...
}
```
It's not the same thing as:
```go
if err != nil && err != http.ErrNoLocation {
// ...
}

if err != http.ErrNoLocation && redirectURL.String() != "" {
// ...
}
```

It's the same as:
```go
if err != nil && err != http.ErrNoLocation {
// ...
}

if err == nil && redirectURL.String() != "" {
// ...
}
```

Are you sure of your changes?

Reply:
Yes, the change is as intended. I believe that because the first if clause in each of the last two snippets you presented above should contain a return statement, the second if clause in both snippets are functionally equivalent (as the only value `err` may contain after the first if clause should be `nil` or `http.ErrNoLocation`).

I believe another way to write this would be with a switch statement:

```go
redirectURL, err := forwardResponse.Location()

switch {

// Do nothing, as there was no Location header.
case err == http.ErrNoLocation:
    break

// Return an error, as we were (likely) unable to parse the Location header.
case err != nil:
    log.Debugf("Error reading response location header %s. Cause: %s", config.Address, err)
    w.WriteHeader(http.StatusInternalServerError)
    return

// Set the location in our response if one was sent back.
case redirectURL.String() != "":
    w.Header().Add("Location", redirectURL.String())

}
```